test/test_common/simulated_time_system.h

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include "envoy/event/timer.h"
+
+#include "common/common/lock_guard.h"
+#include "common/common/thread.h"
+#include "common/common/utility.h"
+
+namespace Envoy {
+namespace Event {
+
+// Represents a simulated time system, where time is advanced by calling
+// sleep(), setSystemTime(), or setMonotonicTime(). systemTime() and
+// monotonicTime() are maintained in the class, and alarms are fired in response
+// to adjustments in time.
+class SimulatedTimeSystem : public TimeSystem {
+public:
+  SimulatedTimeSystem();
+
+  // TimeSystem
+  SchedulerPtr createScheduler(Libevent::BasePtr&) override;
+
+  // TimeSource
+  SystemTime systemTime() override;
+  MonotonicTime monotonicTime() override;
+
+  /**
+   * Advances time forward by the specified duration, running any timers
+   * along the way that have been scheduled to fire.
+   *
+   * @param duration The amount of time to sleep, expressed in any type that
+   * can be duration_casted to MonotonicTime::duration.
+   */
+  template <class Duration> void sleep(const Duration& duration) {
+    mutex_.lock();
+    MonotonicTime monotonic_time =
+        monotonic_time_ + std::chrono::duration_cast<MonotonicTime::duration>(duration);
+    setMonotonicTimeAndUnlock(monotonic_time);
+  }
+
+  /**
+   * Sets the time forward monotonically. If the supplied argument moves
+   * backward in time, the call is a no-op. If the supplied argument moves
+   * forward, any applicable timers are fired, and system-time is also moved
+   * forward by the same delta.
+   *
+   * @param monotonic_time The desired new current time.
+   */
+  void setMonotonicTime(const MonotonicTime& monotonic_time) {
+    mutex_.lock();
+    setMonotonicTimeAndUnlock(monotonic_time);
+  }
+
+  /**
+   * Sets the system-time, whether forward or backward. If time moves forward,
+   * applicable timers are fired and monotonic time is also increased by the
+   * same delta.
+   *
+   * @param system_time The desired new system time.
+   */
+  void setSystemTime(const SystemTime& system_time);
+
+private:
+  class SimulatedScheduler;
+  friend class SimulatedScheduler;
+  class Alarm;
+  friend class Alarm;
+  struct CompareAlarms {
+    bool operator()(const Alarm* a, const Alarm* b) const;
+  };
+  typedef std::set<Alarm*, CompareAlarms> AlarmSet;
+
+  void setMonotonicTimeAndUnlock(const MonotonicTime& monotonic_time) UNLOCK_FUNCTION(mutex_);
```

Comment:
This function also changes system time, I wonder if we should rename.

Reply:
Do you have a suggestion? Above I have the same issue 2x...maybe just a comment suffices?

Reply:
Yeah, it is kind of funny because the type of time in the function name is not what you are setting, but what you are setting both times with... I guess in my mind `setX` is a function that sets X not a function that uses X to set other things. 

Sorry, don't want to split too many hairs, so perhaps just a comment like the above suffices as you suggest.